text/chapter8.md

```diff
@@ -503,124 +510,163 @@ The `ST` effect is used to manipulate mutable state. As pure functional programm
 
 ```javascript
-var ref = { x: x0, v: v0 };
-
-Control_Monad_Eff.forE(0)(time * 1000 | 0)(function (i) {
-  return function __do() {
-    ref = (function (o) {
-      return {
-        v: o.v - 9.81 * 1.0e-3,
-        x: o.x + o.v * 1.0e-3
-      };
-    })(ref);
-    return Prelude.unit;
-  };
-})();
-
-return ref.x;
-```
-
-The `ST` effect is a good way to generate short JavaScript when working with locally-scoped mutable state, especially when used together with actions like `forE`, `foreachE`, `whileE` and `untilE` which generate efficient loops in the `Eff` monad.
+var simulate = function (x0) {
+    return function (v0) {
+        return function (time) {
+            return (function __do() {
+                var ref = {
+                    value: {
+                        x: x0,
+                        v: v0
+                    }
+                };
+                Control_Monad_ST_Internal["for"](0)(time * 1000 | 0)(function (v) {
+                    return Control_Monad_ST_Internal.modify(function (o) {
+                        return {
+                            v: o.v - 9.81 * 1.0e-3,
+                            x: o.x + o.v * 1.0e-3
+                        };
+                    })(ref);
+                })();
+                return ref.value.x;
+            })();
+        };
+    };
+};
+```
+
+For comparison, this is the generated JavaScript of the non-inlined form:
+
+```js
+var simulate = function (x0) {
+    return function (v0) {
+        return function (time) {
+            return function __do() {
+                var ref = Control_Monad_ST_Internal["new"]({
+                    x: x0,
+                    v: v0
+                })();
+                Control_Monad_ST_Internal["for"](0)(time * 1000 | 0)(function (v) {
+                    return Control_Monad_ST_Internal.modify(function (o) {
+                        return {
+                            v: o.v - 9.81 * 1.0e-3,
+                            x: o.x + o.v * 1.0e-3
+                        };
+                    })(ref);
+                })();
+                var $$final = Control_Monad_ST_Internal.read(ref)();
+                return $$final.x;
+            };
+        };
+    };
+};
+```
+
+The `ST` effect is a good way to generate short JavaScript when working with locally-scoped mutable state, especially when used together with actions like `for`, `foreach`, and `while` which generate efficient loops.
+
+## Exercises
+
+1. (Medium) Rewrite the `safeDivide` function to throw an exception using `throwException` if the denominator is zero.
+1. (Difficult) The following is a simple way to estimate pi: randomly choose a large number `N` of points in the unit square, and count the number `n` which lie in the inscribed circle. An estimate for pi is `4n/N`. Use the `random` and `ST` effects with the `for` function to write a function which estimates pi in this way.
```

Comment:
Is it still possible to blend `Effect.random` with `ST`? This is one of the more interesting exercises, so I'd like to keep it. Maybe it should be moved to ch11 in the "State Monad" section.

Reply:
"Once you're in `Effect` you might as well use `Ref`. When we were still using `Eff` you could eliminate the `ST` effect, but that was always kind of dodgy"

Reply:
If the second exercise can't include randomness, then it seems like it could actually be [tackled in ch4](https://book.purescript.org/chapter4.html#exercises-2) like so:
```hs
estimatePi :: Int -> Number
estimatePi r =
  let
    bigN = r * r

    n =
      sum do
        x <- 1 .. r
        y <- 1 .. r
        guard $ x * x + y * y < r * r
        pure 1
  in
    (toNumber (4 * n)) / (toNumber bigN)
```
I'm not sure if adding this to ch4 would be an improvement, since it doesn't test any new language concepts.

There's still the question of what exercise to use for `ST` in this chapter.